REMEMBER: FIRST CREATE A COPY OF THIS FILE WITH A UNIQUE NAME AND DO YOUR WORK THERE. AND MAKE SURE YOU COMMIT YOUR CHANGES TO THE `hw3_submissions` BRANCH.

# Assignment 3 | Cleaning and Exploring Data with Pandas



<img src="data/scoreCard.jpg" width=250>

In this assignment, you will investigate restaurant food safety scores for restaurants in San Francisco. Above is a sample score card for a restaurant. The scores and violation information have been made available by the San Francisco Department of Public Health. 

## Loading Food Safety Data


There are 2 files in the data directory:
1. business.csv containing food establishments in San Francisco
1. inspections.csv containing retaurant inspections records

Let's start by loading them into Pandas dataframes.  One of the files, business.csv, has encoding (ISO-8859-1), so you will need to account for that when reading it.

### Question 1

#### Question 1a
Read the two files noted above into two pandas dataframes named `bus` and `ins`, respectively. Print the first 5 rows of each to inspect them.


In [2]:
import pandas as pd

In [3]:
bus = pd.read_csv('data/businesses.csv', encoding='ISO-8859-1')
ins = pd.read_csv('data/inspections.csv')

In [4]:
bus.head()

business_id                                name  \
0           10                    TIRAMISU KITCHEN   
1           19               NRGIZE LIFESTYLE CAFE   
2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3           31      NORMAN'S ICE CREAM AND FREEZES   
4           45                 CHARLIE'S DELI CAFE   

                         address           city state postal_code   latitude  \
0                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
1   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
2  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
3           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   
4                3202 FOLSOM ST   San Francisco    CA       94110  37.747114   

    longitude  phone_number  
0 -122.403816  +14154217044  
1 -122.421547  +14157763262  
2 -122.403135  +14156779494  
3 -122.419004           NaN  
4 -122.413641  +14156415051

In [5]:
ins.head()

business_id  score      date     type
0           10     82  20160503  routine
1           10     94  20140729  routine
2           10     92  20140114  routine
3           19     94  20160513  routine
4           19     94  20141110  routine

## Examining the Business data

From its name alone, we expect the `businesses.csv` file to contain information about the restaurants. Let's investigate this dataset.

### Question 2

#### Question 2a: How many records are there?

In [6]:
#using shapeattribute to find number of records 

num_records = bus.shape[0]

print("Number of records in the 'bus' DataFrame:", num_records)

Number of records in the 'bus' DataFrame: 6315


#### Question 2b: How many unique business IDs are there?  

In [9]:
num_unique_ids = bus['business_id'].nunique()

print("Number of unique business IDs in the 'bus' DataFrame:", num_unique_ids)

Number of unique business IDs in the 'bus' DataFrame: 6315


#### Question 2c: What are the 5 most common businesses by name, and how many are there in San Francisco?

In [10]:
sf_bus = bus[bus['city'] == 'San Francisco']

In [11]:
top_5_common = sf_bus['name'].value_counts().head()

In [12]:
print("The 5 most common businesses by name in San Francisco:")
print(top_5_common)

The 5 most common businesses by name in San Francisco:
name
STARBUCKS COFFEE              72
PEET'S COFFEE & TEA           24
MCDONALDS                     12
SAN FRANCISCO SOUP COMPANY    11
WALGREENS                     11
Name: count, dtype: int64


## Zip code

Next, let's  explore some of the variables in the business table. We begin by examining the postal code.

### Question 3

#### Question 3a
How are the zip code values stored in python (i.e. data type)?

To answer this you might want to examine a particular entry.

In [13]:
sample_zip_code = bus.loc[0, 'postal_code']

In [14]:
zip_code_data_type = type(sample_zip_code)

In [15]:
print("Data type of zip code values:", zip_code_data_type)

Data type of zip code values: <class 'str'>


#### Question 3b

What are the unique values of postal_code?

In [16]:
unique_zip_codes = bus['postal_code'].unique()

In [17]:
print("Unique values of postal_code:", unique_zip_codes)

Unique values of postal_code: ['94104' '94109' '94133' '94110' '94122' '94103' '94115' '94131' '94111'
 '94117' '94107' '94108' '94102' '94132' '94105' '94134' '94124' '94116'
 '94121' '94118' '94112' '94123' '94114' '94127' '941102019' '941' '94545'
 '94066' '941033148' nan '94158' '95105' '94013' '94130' 'CA' '92672'
 '94120' '94143' '94609' '94101' '00000' '94188' '94621' '94014' '94129'
 '94602' 'Ca']


#### Question 3c

Let's say we decide to exclude the businesses that have no zipcode for our analysis (which might include food trucks for example).  Use the list of valid 5-digit zip codes below to create a new dataframe called bus_valid, with only businesses whose postal_codes show up in this list of valid zipcodes. How many businesses are there in this new dataframe?

In [18]:
validZip = ["94102", "94103", "94104", "94105", "94107", "94108",
            "94109", "94110", "94111", "94112", "94114", "94115",
            "94116", "94117", "94118", "94121", "94122", "94123", 
            "94124", "94127", "94131", "94132", "94133", "94134"]

In [19]:
vb = bus[bus['postal_code'].isin(validZip)]

In [ ]:
#to filter the original dataframe ('bus')to only include the rows where 'postal_code' column matches any of the valid zip codes provided, 'isin()' method will be used to check list of valid zip codes. 

In [20]:
#list of valid 5-digit zip codes
validZip = ["94102", "94103", "94104", "94105", "94107", "94108",
            "94109", "94110", "94111", "94112", "94114", "94115",
            "94116", "94117", "94118", "94121", "94122", "94123", 
            "94124", "94127", "94131", "94132", "94133", "94134"]

In [21]:
bus_valid = bus[bus['postal_code'].isin(validZip)]

In [23]:
num_bus_valid = bus_valid.shape[0]
print("Number of businesses with valid zip codes in the new DataFrame 'bus_valid':", num_bus_valid)


Number of businesses with valid zip codes in the new DataFrame 'bus_valid': 5999


## Latitude and Longitude

Another aspect of the data we want to consider is the prevalence of missing values. If many records have missing values then we might be concerned about whether the nonmissing values are representative of the population.

### Question 4
 
Consider the longitude and latitude in the business DataFrame. 

#### Question 4a

How many businesses are missing longitude values, working with only the businesses that are in the list of valid zipcodes?

In [24]:
vb[pd.isnull(vb['longitude'])]

business_id                               name             address  \
200           821                       LA SANTANECA    3781 MISSION ST    
210           852  NEW HOLLYWOOD BAKERY & RESTAURANT    652 PACIFIC AVE    
760          2934                   LOU'S FISH SHACK   300 JEFFERSON ST    
826          3123                        LES JOULINS        44 ELLIS ST    
1100         4323                             DIVA'S       1081 POST ST    
...           ...                                ...                 ...   
6310        90234               ACCESS SFUSD THE ARC     1520 HOWARD ST    
6311        90236                     PHO DONG HUONG  667 MONTEREY BLVD    
6312        90237                   WIN'S RESTAURANT    3040 TARAVAL ST    
6313        90238                   USA POWER MARKET     1524 OCEAN AVE    
6314        90239                         ALBA RAY'S    2293 MISSION ST    

               city state postal_code  latitude  longitude  phone_number  
200   San Francisco    CA       94110       NaN        NaN  +14156481034  
210   San Francisco    CA       94133       NaN        NaN  +14153979919  
760   San Francisco    CA       94133       NaN        NaN  +14157715687  
826   San Francisco    CA       94102       NaN        NaN  +14153975397  
1100  San Francisco    CA       94109       NaN        NaN  +14154743482  
...             ...   ...         ...       ...        ...           ...  
6310  San Francisco    CA       94103       NaN        NaN           NaN  
6311  San Francisco    CA       94127       NaN        NaN  +14154945048  
6312  San Francisco    CA       94116       NaN        NaN  +14158606036  
6313  San Francisco    CA       94112       NaN        NaN  +14153331878  
6314  San Francisco    CA       94110       NaN        NaN  +14158881393  

[2483 rows x 9 columns]

#### Question 4b

Create a new dataframe with one row for each valid zipcode. The dataframe should include the following three columns:

1. `postal_code`: Contains the zip codes in the `validZip` variable above.
2. `null_lon`: The number of businesses in that zipcode with missing `longitude` values.
3. `not_null_lon`: The number of businesses without missing `longitude` values.

In [25]:
validZip = ["94102", "94103", "94104", "94105", "94107", "94108",
            "94109", "94110", "94111", "94112", "94114", "94115",
            "94116", "94117", "94118", "94121", "94122", "94123", 
            "94124", "94127", "94131", "94132", "94133", "94134"]

In [27]:
vb = bus[bus['postal_code'].isin(validZip)]

In [28]:
null_lon_counts = vb[vb['longitude'].isnull()].groupby('postal_code').size().reset_index(name='null_lon')


In [29]:
not_null_lon_counts = vb[vb['longitude'].notnull()].groupby('postal_code').size().reset_index(name='not_null_lon')

In [30]:
zip_lon_counts = pd.merge(null_lon_counts, not_null_lon_counts, on='postal_code', how='outer').fillna(0)

In [31]:
zip_lon_counts.set_index('postal_code', inplace=True)

In [32]:
print("DataFrame with counts of missing and non-missing longitude values for each valid zipcode:")
print(zip_lon_counts)


DataFrame with counts of missing and non-missing longitude values for each valid zipcode:
             null_lon  not_null_lon
postal_code                        
94102             196           262
94103             252           306
94104              59            74
94105             100           126
94107             247           204
94108              82           142
94109             146           237
94110             251           334
94111             115           167
94112              59           131
94114             108           122
94115              76           162
94116              35            58
94117              67           130
94118             100           143
94121              55           106
94122             123           147
94123              53           116
94124              88            76
94127              28            43
94131              14            34
94132              65            69
94133             135           282
94134     

#### 4c. Do any zip codes appear to have more than their 'fair share' of missing longitude? 

To answer this, you will want to compute the proportion of missing longitude values for each zip code, and print the proportion missing longitude, and print the top five zipcodes in descending order of proportion missing postal_code.


In [18]:
validZip = ["94102", "94103", "94104", "94105", "94107", "94108",
            "94109", "94110", "94111", "94112", "94114", "94115",
            "94116", "94117", "94118", "94121", "94122", "94123", 
            "94124", "94127", "94131", "94132", "94133", "94134"]

vb = bus[bus['postal_code'].isin(validZip)]


# counts of missing and non-missing longitude values for each valid zipcode
null_lon_counts = vb[vb['longitude'].isnull()].groupby('postal_code').size().reset_index(name='null_lon')
not_null_lon_counts = vb[vb['longitude'].notnull()].groupby('postal_code').size().reset_index(name='not_null_lon')

# Merging the counts into a single dataframe
zip_lon_counts = pd.merge(null_lon_counts, not_null_lon_counts, on='postal_code', how='outer').fillna(0)
zip_lon_counts.set_index('postal_code', inplace=True)

#total number of businesses for each zip code
zip_lon_counts['total_businesses'] = zip_lon_counts['null_lon'] + zip_lon_counts['not_null_lon']

#proportion of missing longitude values for each zip code
zip_lon_counts['proportion_missing'] = zip_lon_counts['null_lon'] / zip_lon_counts['total_businesses']

# zip codes based on the proportion of missing longitude values
sorted_zip_lon_counts = zip_lon_counts.sort_values(by='proportion_missing', ascending=False)


print("Top five zip codes with the highest proportion of missing longitude values:")
print(sorted_zip_lon_counts.head())

Top five zip codes with the highest proportion of missing longitude values:
             null_lon  not_null_lon  total_businesses  proportion_missing
postal_code                                                              
94107             247           204               451            0.547672
94124              88            76               164            0.536585
94132              65            69               134            0.485075
94114             108           122               230            0.469565
94122             123           147               270            0.455556


# Investigate the inspection data

Let's now turn to the inspection DataFrame. Earlier, we found that `ins` has 4 columns, these are named `business_id`, `score`, `date` and `type`.  In this section, we determine the granularity of `ins` and investigate the kinds of information provided for the inspections. 

### Question 5

#### Question 5a
As with the business data, assess whether there is one inspection record for each business, by counting how many rows are in the data and how many unique businesses there are in the data. If they are exactly the same number, it means there is only one inspection per business, clearly.

In [24]:
num_inspection_records = ins.shape[0]

# number of unique businesses in the inspection data
num_unique_businesses = ins['business_id'].nunique()

print("Number of inspection records:", num_inspection_records)
print("Number of unique businesses:", num_unique_businesses)

# To verify number of inspection records is equal to the number of unique businesses
if num_inspection_records == num_unique_businesses:
    print("There is one inspection record for each business.")
else:
    print("There is more than one inspection record for some businesses.")

Number of inspection records: 15430
Number of unique businesses: 5730
There is more than one inspection record for some businesses.


#### Question 5b

What values does `type` take on? How many occurrences of each value is in the DataFrame? Create a new dataframe named `ins2` by copying `ins` and keeping only records with values of `type` that occur more than 10 times in the original table. In other words, eliminate records that have values of `type` that occur rarely (< 10 times). Check the result to make sure rare types are eliminated.

In [28]:
# Unique values in "type" column 
type_counts = ins['type'].value_counts()


print("Occurrences of each value in the 'type' column:")
print(type_counts)

# To filter out records with values of "type" that occur rarely (< 10 times)
rare_types = type_counts[type_counts < 10].index
ins2 = ins[~ins['type'].isin(rare_types)]

# Verifying if rare types are eliminated
print("\nNumber of unique values in 'type' column after filtering:", ins2['type'].nunique())
print("Occurrences of each value in the 'type' column after filtering:")
print(ins2['type'].value_counts())


Occurrences of each value in the 'type' column:
type
routine      15429
complaint        1
Name: count, dtype: int64

Number of unique values in 'type' column after filtering: 1
Occurrences of each value in the 'type' column after filtering:
type
routine    15429
Name: count, dtype: int64


#### Question 5c

Since the data was stored in a .csv file, the dates are formatted as strings such as `20160503`. Once we read in the data, we would like to have dates in an appropriate format for analysis. Add a new column called `year` by capturing the first four characters of the date column. 

Hint: we have seen multiple ways of doing this in class, includings `str` operations, `lambda` functions, `datetime` operations, and others. Choose the method that works best for you :)

In [39]:
# Converting 'date' column to string type
ins['date'] = ins['date'].astype(str)

# New column 'year'to capture the first four characters of the 'date' column
ins['year'] = ins['date'].str[:4]

print(ins.head())

   business_id  score      date     type  year
0           10     82  20160503  routine  2016
1           10     94  20140729  routine  2014
2           10     92  20140114  routine  2014
3           19     94  20160513  routine  2016
4           19     94  20141110  routine  2014


#### Question 5d

What range of years is covered in this data set? Are there roughly same number of inspections each year? Try dropping records for any years with less than 50 inspections and store the result in a new dataframe named `ins3`.

In [41]:
# looking for range of years covered in the dataset
min_year = ins['year'].min()
max_year = ins['year'].max()
print("Range of years covered:", min_year, "-", max_year)

# number of inspections each year
inspection_counts_per_year = ins['year'].value_counts().sort_index()
print("\nNumber of inspections each year:\n", inspection_counts_per_year)

# Droping records for any years with less than 50 inspections
ins3 = ins.groupby('year').filter(lambda x: len(x) >= 50)

# Display the updated DataFrame with the dropped records
print("\nDataFrame 'ins3' after dropping records for years with less than 50 inspections:")
print(ins3.head())

Range of years covered: 2013 - 2016

Number of inspections each year:
 year
2013      38
2014    5629
2015    3923
2016    5840
Name: count, dtype: int64

DataFrame 'ins3' after dropping records for years with less than 50 inspections:
   business_id  score      date     type  year
0           10     82  20160503  routine  2016
1           10     94  20140729  routine  2014
2           10     92  20140114  routine  2014
3           19     94  20160513  routine  2016
4           19     94  20141110  routine  2014


Let's examine only the inspections for one year: 2016. This puts businesses on a more equal footing because [inspection guidelines](https://www.sfdph.org/dph/eh/Food/Inspections.asp) generally refer to how many inspections should occur in a given year.

In [42]:
# Filtering inspections for the year 2016
ins_2016 = ins[ins['year'] == '2016']


print(ins_2016.head())

    business_id  score      date     type  year
0            10     82  20160503  routine  2016
3            19     94  20160513  routine  2016
6            24     98  20161005  routine  2016
7            24     96  20160311  routine  2016
10           31     84  20160816  routine  2016


### Question 6

#### Question 6a

Merge the business and 2016 inspections data, keeping all businesses regardless of whether they show up in the inspections file. Show the first several rows of the resulting dataframe.

In [46]:
# To merge business and 2016 inspections data
bus_ins_2016 = pd.merge(bus, ins_2016, on='business_id', how='left')

#pd.merge() function to dataframs


print(bus_ins_2016.head())


   business_id                                name  \
0           10                    TIRAMISU KITCHEN   
1           19               NRGIZE LIFESTYLE CAFE   
2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
4           31      NORMAN'S ICE CREAM AND FREEZES   

                         address           city state postal_code   latitude  \
0                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
1   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
2  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
3  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
4           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   

    longitude  phone_number  score      date     type  year  
0 -122.403816  +14154217044   82.0  20160503  routine  2016  
1 -122.421547  +14157763262   94.0  20160513  routine  2016  
2 -122.4

#### Question 6b
Print the 20 lowest rated businesses names, their addresses, and their ratings.

In [47]:
lowest_rated = bus_ins_2016[['name', 'address', 'score']].sort_values(by='score').head(20)
print(lowest_rated)

                             name              address  score
4892      GOLDEN RIVER RESTAURANT     5827 GEARY BLVD    52.0
7325                    POKI TIME     2101 LOMBARD ST    55.0
6281                 CRAZY PEPPER   2257 SAN JOSE AVE    55.0
4873                   UNCLE CAFE       65 WAVERLY PL    55.0
3008             HING WANG BAKERY        339 JUDAH ST    56.0
5627                   GOLDEN WOK   295 B ORIZABA AVE    56.0
5383                   TOP'S CAFE  2460 SAN BRUNO AVE    57.0
6067    L & G VIETNAMESE SANDWICH         602 EDDY ST    57.0
836                  RED A BAKERY      634 CLEMENT ST    57.0
4209      SAN TUNG RESTAURANT LLC      1031 IRVING ST    57.0
549           MA'S DIM SUM & CAFE      1315 POWELL ST    57.0
4903  NEW GARDEN RESTAURANT, INC.       716 KEARNY ST    57.0
6645    KOBANI MEDITERANEAN GRILL       3242 SCOTT ST    58.0
624           JOANIE'S DINER INC.   1329 COLUMBUS AVE    59.0
6938         PAK NAM THAI CUISINE       655 LARKIN ST    59.0
3563    

## Done!

Now commit this notebook to your `hw3_submissions` branch, push it to your GitHub repo, and open a PR!